security/oc_tls.c

```diff
@@ -1342,32 +1379,74 @@ oc_tls_configure_end_entity_cert_chain(mbedtls_ssl_config *conf, size_t device,
   oc_x509_crt_t *cert = (oc_x509_crt_t *)oc_list_head(g_identity_certs);
   while (cert != NULL) {
     if (cert->device == device && cert->cred->credusage == credusage &&
-        (credid == -1 || cert->cred->credid == credid)) {
+        (credid == OC_TLS_SELECTED_ANY_CRED_ID ||
+         cert->cred->credid == credid)) {
       break;
     }
     cert = cert->next;
   }
-
-  if (!cert || mbedtls_ssl_conf_own_cert(conf, &cert->cert, &cert->pk) != 0) {
-    OC_WRN("error configuring identity cert");
+  if (!cert) {
+#if OC_WRN_IS_ENABLED
+    char credid_str[16];
+    memset(credid_str, 0, sizeof(credid_str));
+    if (credid == OC_TLS_SELECTED_ANY_CRED_ID) {
+      strncpy(credid_str, "any", sizeof(credid_str));
+    } else {
+      snprintf(credid_str, sizeof(credid_str), "%d", credid);
+    }
+    OC_WRN(
+      "cannot set client %s certificate(selected %s): certificate not found",
+      credusage == OC_CREDUSAGE_MFG_CERT ? "manufacturer" : "identity",
+      credid_str);
+#endif /* OC_WRN_IS_ENABLED */
+    return -1;
+  }
+  int err = mbedtls_ssl_conf_own_cert(conf, &cert->cert, &cert->pk);
+  if (err != 0) {
+#if OC_WRN_IS_ENABLED
+    char credid_str[16];
+    memset(credid_str, 0, sizeof(credid_str));
+    if (credid == OC_TLS_SELECTED_ANY_CRED_ID) {
+      strncpy(credid_str, "any", sizeof(credid_str));
+    } else {
+      snprintf(credid_str, sizeof(credid_str), "%d", credid);
+    }
+    char buf[128];
+    mbedtls_strerror(err, buf, sizeof(buf));
+    OC_WRN("cannot set client %s certificate(selected %s): %s",
+           credusage == OC_CREDUSAGE_MFG_CERT ? "manufacturer" : "identity",
+           credid_str, buf);
+#endif /* OC_WRN_IS_ENABLED */
     return -1;
   }
 
   return 0;
 }
 
-static int
-oc_tls_load_mfg_cert_chain(mbedtls_ssl_config *conf, size_t device, int credid)
+#ifndef OC_TEST
+static
+#endif /* !OC_TEST */
+  int
+  oc_tls_load_mfg_cert_chain(mbedtls_ssl_config *conf, size_t device,
+                             int credid)
 {
   OC_DBG("loading manufacturer cert chain");
   return oc_tls_configure_end_entity_cert_chain(conf, device,
                                                 OC_CREDUSAGE_MFG_CERT, credid);
 }
 
-static int
-oc_tls_load_identity_cert_chain(mbedtls_ssl_config *conf, size_t device,
-                                int credid)
+#ifndef OC_TEST
```

Comment:
Just put the declaration outside the `OC_TEST` ifdef in the header. `oc_tls_validate_identity_certs_consistency` and `oc_tls_validate_trust_anchors_consistency` declarations are ifdef-ed in the header because the definitions exist only with `OC_TEST`, otherwise internal functions have declarations in internal headers to be available for tests.

Reply:
Fixed